Add support of date time ranges. Time range support for download_backtest_data.

arudov opened the pull request:
## Summary
Add support of downloading test data for a particular number of days

Solve the issue: #607

## Quick changelog

- add support of number of days
- exchange name could be pass as the parameter
- add "date" time-ranges support

xmatthias commented on scripts/download_backtest_data.py:
```diff
@@ -27,15 +30,36 @@
+
+        # download data until it reaches today now time
+        data = []
+        while not since_time or since_time < (time.time() - 10 * MIN_SECCONDS) * 1000:
```
I think this loop should be moved into `get_ticker_history` - so it can be used by the backtesting-script as well (when using --refresh-pairs-cached).

xmatthias commented on scripts/download_backtest_data.py:
```diff
@@ -4,8 +4,8 @@
 import sys
 import os
-
-from freqtrade import (exchange, arguments, misc)
```
I think you'll want to keep this import, otherwise `exchange.` in line 34 and `arguments.` in line 12 will not be available.

xmatthias commented on scripts/download_backtest_data.py:
```diff
@@ -27,15 +30,36 @@
+            data_part = exchange.get_ticker_history(pair, tick_interval, since=since_time)
+
+            if not data_part:
+                print('\tNo data since %s' % datetime.datetime.utcfromtimestamp(since_time / 1000).strftime('%Y-%m-%dT%H:%M:%S'))
```
I know flake8 is not checked in the scripts-directory ... but i still think this line is too long and we should stick to the 100 char limit

xmatthias commented on scripts/download_backtest_data.py:
```diff
@@ -27,15 +30,36 @@
+            
+            print('\tData received for period %s - %s' % 
+                    (datetime.datetime.utcfromtimestamp(data_part[0][0] / 1000).strftime('%Y-%m-%dT%H:%M:%S'),
+                    datetime.datetime.utcfromtimestamp(data_part[-1][0] / 1000).strftime('%Y-%m-%dT%H:%M:%S')))
```
linelength - same as above

xmatthias commented on scripts/download_backtest_data.py:
```diff
@@ -27,15 +30,36 @@
+
+        since_time = None
+        if args.days:
+            since_time = int((time.time() - args.days * DAY_SECCONDS) * 1000)
```
could be simplified to `arrow.utcnow().shift(days=-args.days).timestamp * 1000` - removing the need for `MIN_SECCONDS` and `DAY_SECCONDS` and `HOUR_SECCONDS`

xmatthias commented on scripts/download_backtest_data.py:
```diff
@@ -15,6 +15,9 @@
 PAIRS = []
+MIN_SECCONDS = 60
+HOUR_SECCONDS = 60 * MIN_SECCONDS
+DAY_SECCONDS = 24 * HOUR_SECCONDS
```
seconds may seem long sometimes - but they did not earn a 2nd C yet ;)

arudov replied:
=))) yeah! Just wrote SEC and then added CONDS=)

Will do all the fixes later today

xmatthias commented on freqtrade/optimize/__init__.py:
```diff
@@ -112,12 +139,31 @@ def download_pairs(datadir, pairs: List[str], ticker_interval: str) -> bool:
+    return None
+
+
 # FIX: 20180110, suggest rename interval to tick_interval
```
not directly related to this PR - but if we touch this function - should we also consider applying this comment?

arudov replied:
will do

arudov commented on freqtrade/exchange/__init__.py:
```diff
@@ -269,9 +269,28 @@ def get_ticker(pair: str, refresh: Optional[bool] = True) -> dict:
+        # it looks as if some exchanges return cached data
+        # and update it with some delay so 10 mins interval is added
+        data = []
+        while not since or since < arrow.utcnow().shift(minutes=-10).timestamp * 1000:
```
I just want to prevent doing the extra call. So 10 minutes is just a delay to adjust a logic for cashes of exchanges API. From my testing it looks like 10 minutes is enough for exchanges to update the data. 

1h+ tickers needs some other checks. Will add them.

arudov replied:
The documentation says that `fetch_ohlcv` will return the data that starts after since time. So there should not be any records duplications.

xmatthias replied:
works for me - just asked as it was unclear to me :+1:

arudov commented on freqtrade/optimize/__init__.py:
```diff
@@ -134,23 +180,33 @@ def download_backtesting_testdata(datadir: str, pair: str, interval: str = '5m')
-    new_data = get_ticker_history(pair=pair, tick_interval=interval)
-    for row in new_data:
-        if row not in data:
-            data.append(row)
```
If there is the existing data `since` parameter will be equal to the `max({requested time by user}, {last exist data time} + 1)` then `get_ticker_history` will return the data w/o duplications and only starting from `since` time. So there could be the duplication only if CCXT library has a bug in it(it could be, like this one https://github.com/ccxt/ccxt/issues/379, but looks like they all were fixed;) ).

I can revert this code if you want. But it will be like a double-cleck of CCXT response.

xmatthias replied:
Makes sense to me (also, i did miss the connection to line 198 - which sets since to the last point downloaded previously).

arudov commented on freqtrade/optimize/__init__.py:
```diff
@@ -134,23 +180,33 @@ def download_backtesting_testdata(datadir: str, pair: str, interval: str = '5m')
+
     logger.debug("New Start: %s", misc.format_ms_time(data[0][0]))
-    logger.debug("New End: %s", misc.format_ms_time(data[-1:][0][0]))
-    data = sorted(data, key=lambda data: data[0])
```
The [CCXT doc](https://github.com/ccxt/ccxt/wiki/Manual#market-data) says `The list of candles is returned sorted in ascending (historical) order, oldest candle first, most recent candle last.` SO I can revert back the code, but it shouldn't be necessary.

PS. I forgot about that note "Note that the info from the last (current) candle may be incomplete until the candle is closed (until the next candle starts)." Will add changes for this case.

arudov commented on freqtrade/tests/optimize/test_optimize.py:
```diff
@@ -145,6 +147,28 @@ def test_download_pairs(ticker_history, mocker) -> None:
+    mocker.patch('arrow.utcnow', return_value=mock_now)
+
+    start = get_start_ts_from_timerange(((None, 'line'), None, -200), '1m')
+    assert start == (1367900664 - 200 * 60) * 1000
```
ok

arudov commented:
@xmatthias I've updated the code. I changed the logic a little bit: now the bot will always download all data that does not exist in cached files ignoring timerange parameter. It is done to be sure that there is no missed data in the center of cached data.
For example if you download data for 1 day and in 5 days ask the script to update the data for 1 day again it will download the data for 5 days. It allows not to have an empty windows between two downloads.

arudov commented:
@gcarq The branch fix/dl-testdata was merged to objectify-ccxt branch, does it make sense to me to create my pull-request to that branch?

gcarq commented:
@arudov Yes please, that would be awesome. I will take a look in a little bit

arudov commented:
Created a new PR https://github.com/gcarq/freqtrade/pull/637
